ygen/protogen.go

```diff
@@ -403,7 +531,19 @@ func genProto3Msg(msg *yangDirectory, msgs map[string]*yangDirectory, state *gen
 			}
 
 			if keyMsg != nil {
-				msgDefs = append(msgDefs, *keyMsg)
+				if cfg.nestedMessages {
+					// If nested messages are being output, we must ensure that the
+					// generated key message is output within the parent message - hence
+					// it is generated directly here and appended to the child messages.
+					kc, cerrs := genProto3MsgCode(parentPkg, []protoMsg{*keyMsg})
```

Comment:
The fact that *keyMsg is being used suggests that perhaps genProto3MsgCode should take a slice of ptr instead of value?